server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -2082,12 +2082,61 @@ void handleBonusPools(Owner owner, Map<String, PoolQuantity> pools,
         @Override
         public void handlePostEntitlement(PoolManager manager, Consumer consumer,
             Map<String, Entitlement> entitlements) {
-            Set<String> stackIds = new HashSet<String>();
+
+            /*
+             * selectBestPools can return multiple pools on the same stack.
+             * Hence there can be more than one entitlement with the same stack id here.
+             * enforcer.postEntitlement may try to create more than one stack derived pool
+             * per stack id if we provide that as input, so we sort the entitlements.
+             *
+             * Entitlements that share stack ids need to be operated individually.
+             * other entitlements can be operated in a batch.
+             *
+             */
+            Map<String, Integer> stackIdCounts = new HashMap<String, Integer>();
             for (Entitlement entitlement : entitlements.values()) {
                 if (entitlement.getPool().isStacked()) {
-                    stackIds.add(entitlement.getPool().getStackId());
+                    String stackId = entitlement.getPool().getStackId();
+                    Integer count = stackIdCounts.containsKey(stackId) ?
+                        stackIdCounts.get(stackId) : 0;
+                    stackIdCounts.put(stackId, count + 1);
                 }
             }
+            Set<String> uniqueStackIds = new HashSet<String>();
+            Map<String, Entitlement> uniqueEnts = new HashMap<String, Entitlement>();
+            Map<String, Entitlement> nonUniqueEnts = new HashMap<String, Entitlement>();
+            for (Entry<String, Entitlement> entry : entitlements.entrySet()) {
+                Pool p = entry.getValue().getPool();
+                if (p.isStacked() && stackIdCounts.get(p.getStackId()) > 1) {
+                    nonUniqueEnts.put(entry.getKey(), entry.getValue());
+                }
+                else {
+                    uniqueEnts.put(entry.getKey(), entry.getValue());
+                    if (p.isStacked()) {
+                        uniqueStackIds.add(p.getStackId());
+                    }
+                }
+            }
+            // ents with unique stack id can be operated in a batch
+            if (!uniqueEnts.isEmpty()) {
+                handlePostEntitlement(manager, consumer, uniqueEnts, uniqueStackIds);
+            }
+            // others must be operated individually
+            for (Entry<String, Entitlement> entry: nonUniqueEnts.entrySet()) {
+                Entitlement ent = entry.getValue();
+                Map<String, Entitlement> entMap = new HashMap<String, Entitlement>();
+                entMap.put(entry.getKey(), ent);
```

Comment:
Tip: you can use Collections.<String, Entitlement>singletonMap(entry.getKey(), ent)) to save some boiler plate and tiny bits of memory and overhead here. singletonMap returns an immutable (!) map of the given type with the mapping passed in. The immutable nature means it can't always be used, but as a result there's no bucket management so everything is slightly faster and easier on resources.